File: src/pnp/compiler/model/Variable.java
package pnp.compiler.model;

import pnp.compiler.model.type.Type;

public class Variable implements Expression {
    final private Type type;
    private Object value;
    final private String name;

    public Variable(Type type) {
        this.type = type;
        this.name = null;
    }

    public Variable(Type type, String name) {
        this.type = type;
        this.name = name;
    }

    public Variable(Type type, Object value) {
        this.type = type;
        this.name = null;
        setValue(value);
    }

    public Variable(Type type, String name, Object value) {
        this.type = type;
        this.name = name;
//        setValue(value);
        this.value = value;
    }

    public Object getValue() {
        return value;
    }

    public void setValue(Object value) {
        if (type.isTypeOf(value)) {
            this.value = value;
        }
    }

    @Override
    public Type getType() {
        return type;
    }

    public String getName() {
        return name;
    }

    public boolean isLiteral() {
        return name == null;
    }
}

File: src/main/kotlin/br/com/pnp/model/construct/Variable.kt
package br.com.pnp.model.construct

import br.com.pnp.model.construct.type.Type
import br.com.pnp.model.construct.type.primitive.PrimitiveType
import br.com.pnp.model.expression.Expression

class Variable(
    override val type: Type,
    override val name: String,
    value: Any? = null
) : Construct, Expression {
    var value: Any? private set

    init {
        value?.let {
            if (!type.isTypeOf(it)) {
                throw IllegalArgumentException(
                        """
                        Type mismatch in ${this.javaClass.simpleName} constructor:
                        Required: $type
                        Found: ${Type.typeOf(it)}
                    """.trimIndent())
            }
        }
        this.value = value
    }

    companion object {
        fun literalInteger(value: Any? = null): Variable {
            return Variable(PrimitiveType.integer, "", value)
        }

        fun literalRational(value: Any? = null): Variable {
            return Variable(PrimitiveType.rational, "", value)
        }

        fun literalBoolean(value: Any? = null): Variable {
            return Variable(PrimitiveType.boolean, "", value)
        }

        fun literalCharacter(value: Any? = null): Variable {
            return Variable(PrimitiveType.character, "", value)
        }

        fun literalString(value: Any? = null): Variable {
            return Variable(PrimitiveType.string, "", value)
        }

        fun literal(type: Type, value: Any? = null): Variable {
            return Variable(type, "", value)
        }
    }

    fun isLiteral(): Boolean {
        return name.isBlank()
    }

    override fun equals(other: Any?): Boolean {
        if (other is Variable) {
            return other.name == this.name && this.type.isTypeOf(other.type)
        }
        return super.equals(other)
    }
}
